### PR TITLE
scheduler_perf: updates to enhance performance testing of DRA

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

These are various enhancements that came out of https://github.com/kubernetes/kubernetes/pull/127277.

#### Special notes for your reviewer:

See also discussions on Slack:
- https://kubernetes.slack.com/archives/C09TP78DV/p1726252355710539
- https://kubernetes.slack.com/archives/C09TP78DV/p1726131630394449

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @macsko

### CODE REVIEW

**Comment by pohly** [comment]
Rebased and removed the unnecessary staging/src/k8s.io/component-base/metrics/testutil/metrics.go change.

**Comment by pohly** [comment]
/hold

To avoid merging with just my self-approval. Feel free to lift the hold if you don't want a second approver.

File: test/integration/scheduler_perf/util.go [hunk 1]
```diff
@@ -396,11 +427,75 @@ func (tc *throughputCollector) init() error {
 }
 
 func (tc *throughputCollector) run(tCtx ktesting.TContext) {
-	podsScheduled, _, err := getScheduledPods(tc.podInformer, tc.namespaces...)
+	// The collector is based on informer cache events instead of periodically listing pods because:
+	// - polling causes more overhead
```
Comment (macsko) [inline test/integration/scheduler_perf/util.go:431]: Is it tested on the fastest workloads? Like `SchedulingBasic`?

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go [hunk 2]
```diff
@@ -433,3 +433,47 @@ func LabelsMatch(metric *dto.Metric, labelFilter map[string]string) bool {
...
+						values[labelPair.GetValue()] = counter.GetValue()
+					}
+				}
+			}
+		}
+	}
+	return values, nil
+}
+
+func findMetric(metricFamilies []*dto.MetricFamily, metricName string) *dto.MetricFamily {
```
[the remark this answers is not in the document]
Comment (pohly) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:472]: Yes.

That `GetHistogramVecFromGatherer` has a unit test reminded me that the new `GetCounterVecFromGatherer` also should have one - added.

File: test/integration/scheduler_perf/gnuplot.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env bash
+
+# Copyright 2023 The Kubernetes Authors.
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/gnuplot.sh:3]: The file really is from 2023 :grin: 

Anyway, udpated.

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 4]
```diff
@@ -373,14 +378,31 @@ func (p *params) UnmarshalJSON(b []byte) error {
...
+	f, err := getParam[float64](p, key)
+	if err != nil {
+		return 0, err
+	}
+	return int(f), nil
+}
+
+// getParam retrieves the parameter as specific type. There is no conversion,
+// so in practice this means that only types that JSON unmarshaling uses
+// (float64, string, bool) work.
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf.go:393]: Good idea. Then it becomes a compile error when used incorrectly.

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 5]
```diff
@@ -591,9 +638,6 @@ type createPodsOp struct {
 }
 
 func (cpo *createPodsOp) isValid(allowParameterization bool) error {
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf.go:640]: Added.

File: test/integration/scheduler_perf/util.go (hunk 1 above)
Comment (pohly) [reply test/integration/scheduler_perf/util.go:431]: I compared the current master (= e456fbfaa64) against the latest revision of this PR (= 0f47ac56907). The short `SchedulingBasic` workloads complete too fast to gather throughput. For the longer ones there is no significant difference:

```
                                                      │     before      │                 after                 │
                                                      │ runtime_seconds │ runtime_seconds  vs base              │
PerfScheduling/SchedulingBasic/5Nodes-36                    12.36 ± 11%       12.34 ± 11%       ~ (p=0.677 n=6)
PerfScheduling/SchedulingBasic/500Nodes-36                  13.92 ±  1%       13.93 ±  1%       ~ (p=0.565 n=6)
PerfScheduling/SchedulingBasic/5000Nodes-36                 28.06 ±  0%       28.11 ±  1%       ~ (p=0.394 n=6)
PerfScheduling/SchedulingBasic/5000Nodes_10000Pods-36       42.52 ±  3%       42.69 ±  3%       ~ (p=0.310 n=6)
geomean                                                     21.29             21.31        +0.09%

...

                                                      │            before            │                       after                        │
                                                      │ SchedulingThroughput/Average │ SchedulingThroughput/Average  vs base              │
PerfScheduling/SchedulingBasic/5000Nodes-36                               397.2 ± 4%                     389.7 ± 6%       ~ (p=0.509 n=6)
PerfScheduling/SchedulingBasic/5000Nodes_10000Pods-36                     609.7 ± 3%                     609.0 ± 4%       ~ (p=0.394 n=6)
geomean                                                                   492.1                          487.1       -1.02%

                                                      │           before            │                       after                       │
                                                      │ SchedulingThroughput/Perc50 │ SchedulingThroughput/Perc50  vs base              │
PerfScheduling/SchedulingBasic/5000Nodes-36                              397.2 ± 4%                    389.7 ± 6%       ~ (p=0.509 n=6)
PerfScheduling/SchedulingBasic/5000Nodes_10000Pods-36                    610.0 ± 2%                    605.0 ± 2%       ~ (p=0.589 n=6)
geomean                                                                  492.3                         485.6       -1.36%

                                                      │           before            │                       after                       │
                                                      │ SchedulingThroughput/Perc90 │ SchedulingThroughput/Perc90  vs base              │
PerfScheduling/SchedulingBasic/5000Nodes-36                              397.2 ± 4%                    389.7 ± 6%       ~ (p=0.509 n=6)
PerfScheduling/SchedulingBasic/5000Nodes_10000Pods-36                    643.4 ± 2%                    639.4 ± 3%       ~ (p=0.699 n=6)
geomean                                                                  505.6                         499.2       -1.26%

                                                      │           before            │                       after                       │
                                                      │ SchedulingThroughput/Perc95 │ SchedulingThroughput/Perc95  vs base              │
PerfScheduling/SchedulingBasic/5000Nodes-36                              397.2 ± 4%                    389.7 ± 6%       ~ (p=0.509 n=6)
PerfScheduling/SchedulingBasic/5000Nodes_10000Pods-36                    804.8 ± 2%                    794.6 ± 3%       ~ (p=0.699 n=6)
geomean                                                                  565.4                         556.5       -1.58%

                                                      │           before            │                       after                       │
                                                      │ SchedulingThroughput/Perc99 │ SchedulingThroughput/Perc99  vs base              │
PerfScheduling/SchedulingBasic/5000Nodes-36                              397.2 ± 4%                    389.7 ± 6%       ~ (p=0.509 n=6)
PerfScheduling/SchedulingBasic/5000Nodes_10000Pods-36                    804.8 ± 2%                    794.6 ± 3%       ~ (p=0.699 n=6)
geomean                                                                  565.4                         556.5       -1.58%
```

Comment (macsko) [reply test/integration/scheduler_perf/util.go:431]: Great. We need to revisit the throughput thresholds some time later, as this change can affect the results a bit.

**Comment by macsko** [comment]
Recently I added `deletePodsOp` to scheduler_perf. Please rebase before merging and handle its opcode appropriately.

**Comment by pohly** [comment]
> Recently I added deletePodsOp to scheduler_perf. Please rebase before merging and handle its opcode appropriately.

That is already included.

**Comment by pohly** [comment]
/assign @rainbowmango

For metrics util approval.

/hold cancel

I think with @macsko in favor we don't need a second SIG Scheduling approver.

**Comment by RainbowMango** [comment]
I will take a look ASAP.

**Comment by pohly** [comment]
I resolved another conflict with new ops.

@macsko: can you double-check and re-add the LGTM?

**Comment by pohly** [comment]
Darn, during resolving conflicts I dropped my own new op :sweat_smile: Will fix tomorrow.

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (RainbowMango) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go:468]: No big deal, just prefer the function name as `findMetricFamily`.

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 6]
```diff
@@ -375,14 +380,31 @@ func (p *params) UnmarshalJSON(b []byte) error {
 	return nil
 }
 
-// get returns param.
+// getParam retrieves the parameter as an integer
```
Comment (macsko) [inline test/integration/scheduler_perf/scheduler_perf.go:383]:
```suggestion
// get retrieves the parameter as an integer
```

File: test/integration/scheduler_perf/scheduler_perf.go [hunk 7]
```diff
@@ -404,43 +426,53 @@ type op struct {
...
+		createPodsOpcode:             &createPodsOp{},
+		createPodSetsOpcode:          &createPodSetsOp{},
+		deletePodsOpcode:             &deletePodsOp{},
+		createResourceClaimsOpcode:   &createResourceClaimsOp{},
+		createResourceDriverOpcode:   &createResourceDriverOp{},
+		churnOpcode:                  &churnOp{},
+		barrierOpcode:                &barrierOp{},
+		sleepOpcode:                  &sleepOp{},
+		startCollectingMetricsOpcode: &startCollectingMetricsOp{},
+		stopCollectingMetricsOpcode:  &stopCollectingMetricsOp{},
```
Comment (macsko) [inline test/integration/scheduler_perf/scheduler_perf.go:443]: I think you forgot to remove opcode checks in collecting metrics ops `isValid` methods.

Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf.go:443]: Fixed.

File: test/integration/scheduler_perf/scheduler_perf.go (hunk 6 above)
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf.go]: Fixed. I had initially experimented with turning this into a generic method, but Go does not support those.

File: staging/src/k8s.io/component-base/metrics/testutil/metrics.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pohly) [reply staging/src/k8s.io/component-base/metrics/testutil/metrics.go]: I missed this when doing my latest update. [Fixed](https://github.com/kubernetes/kubernetes/compare/6f8dea61a9da84a6324dab1fdb532ba6106dfa3b..d100768d940ced902307b29907b11d5c0f63c1bc).

@RainbowMango: can you re-add the LGTM and approve?

**Review by RainbowMango — APPROVED** [review]
/approve
for metrics testutil.
I didn't go through the whole background of this PR, but looks good to have another util to get counter values from gather.

**Pull request merged**